tensorflow/compiler/jit/cluster_scoping_pass.cc

```diff
@@ -0,0 +1,152 @@
+  return absl::nullopt;
+}
+
+void SetXlaScope(Node* node, StringPiece scope) {
+  node->AddAttr(kXlaScopeAttr, scope);
+}
+
+// NB! We append new scope as suffix to the XlaScope attribute instead of
+// overriding the old value.  In this way, we respect the original scopes.
+// In other words, appending X to Y creates the conjunction of the scopes X
+// and Y (i.e, X & Y in effect).
+void AddOrAppendScope(Node* node, absl::string_view suffix) {
+  string updated_scope;
+  absl::optional<string> cur_scope = GetXlaScope(node);
+  if (cur_scope == absl::nullopt) {
+    updated_scope = std::string(suffix);
+  } else {
+    updated_scope = absl::StrCat(cur_scope.value(), "&", suffix);
+  }
+  SetXlaScope(node, updated_scope);
+}
+
+void ClusterScopingPassImpl::AddScopeToAllPredecessors(Node* start) {
+  const string unique_suffix = absl::StrCat("_", GetUniqueScopeId());
+
+  std::vector<Node*> starts;
+  starts.push_back(start);
+  auto enter = [&](Node* n) { AddOrAppendScope(n, unique_suffix); };
+  ReverseDFSFrom(*graph_, starts, enter, /*leave=*/nullptr,
+                 /*stable_comparator=*/NodeComparatorName());
+}
+
+void ClusterScopingPassImpl::AddScopeToAllSuccessors(Node* start) {
+  const string unique_suffix = absl::StrCat("_", GetUniqueScopeId());
+
+  std::vector<Node*> starts;
+  starts.push_back(start);
+  auto enter = [&](Node* n) { AddOrAppendScope(n, unique_suffix); };
+  auto not_back_edge = [](const Edge& edge) -> bool {
+    return !edge.src()->IsNextIteration();
+  };
+  DFSFrom(*graph_, starts, enter, /*leave=*/nullptr,
+          /*stable_comparator=*/NodeComparatorName(),
+          /*edge_filter=*/not_back_edge);
+}
+
+Status ClusterScopingPassImpl::ScopingForPipelineStages() {
+  for (Node* n : graph_->nodes()) {
+    DCHECK(n);
+    if (n->type_string() == "Unstage") {
+      AddScopeToAllSuccessors(n);
+    }
+    if (n->type_string() == "Stage") {
+      AddScopeToAllPredecessors(n);
+    }
+  }
+
+  return Status::OK();
+}
+
+Status ClusterScopingPassImpl::Run() {
+  if (global_jit_level_ == OptimizerOptions::OFF) {
+    return Status::OK();
+  }
+
+  // This preserves the parallelism between pipeline stages.  For example,
+  // below is a typical pattern of input pipelining in Tensorflow and this
+  // heuristic ensures Node_X and Node_Y are put into different clusters.
+  // Without the heuristic, they may be put into the same cluster and it
+  // can introduce artificial dependencies and incur great performance loss.
+  // In this example, Node_Y becomes dependent on IteratorGetNext and the
+  // latencies add up.
+  //
+  // IteratorGetNext -> Node_X -> Stage
+  //
+  // Unstage -> Node_Y
+  //
+  TF_RETURN_IF_ERROR(ScopingForPipelineStages());
+
+  return Status::OK();
+}
+}  // namespace
+
+Status ClusterScopingPass::Run(const GraphOptimizationPassOptions& options) {
+  Graph* graph = options.graph->get();
+
+  TF_RETURN_IF_ERROR(
+      ClusterScopingPassImpl{graph, GetGlobalJitLevelForGraph(options)}.Run());
```

Comment:
This can be just `return ClusterScopingPassImpl{graph, GetGlobalJitLevelForGraph(options)}.Run();`.